### ISSUE TITLE
Dynamically determine device order for t3k

### ISSUE THREAD

**Issue by ttmchiou** [tenstorrent/tt-metal#10608]
**Is your feature request related to a problem? Please describe.**
attempting to setup t3k VMs but I'm getting Device X is not connected to Device Y errors. 
Functionally VM devices are networked together in the correct topology but are not in a expected enumeration order.

We have `t3k_device_mesh ` and `get_devices_for_t3000` which are hardcoding the device enumeration path which throws an error in our CI pipelines since T3K VMs are not in the exact order as hardcoded

**Describe the solution you'd like**
Dynamically determine correct enumeration path instead of hard coding path to allow for t3k VMs to work on t3ks

**Additional context**
Fixing this issue should unblock us from adding from T3K VMs to be added to metal CI pool and reduce the large load on t3k pipelines

**Comment by TT-billteng** [comment]
Bumping priority, we should get this fixed

**Comment by aliuTT** [comment]
I'll take a look and get a PR up.

**Comment by aliuTT** [comment]
I have a WIP PR for this feature: https://github.com/tenstorrent/tt-metal/pull/11485/files#diff-a1cdedf225791630ee0943e20993c96254d86d5822df40d7b4f304e44aa1ebfa

I've added it to DeviceMesh, so maybe we can port the conftest t3k fixtures to use ttnn's `open_device_mesh`?

**Comment by TT-billteng** [comment]
@cfjchu can you help getting this feature in? We really need to add more function T3K machines into CI after discussing with @uaydonat and @jvasilje

**Comment by TT-billteng** [comment]
I hear the fix that @aliuTT is putting in is inferring properties of the topology that syseng already has knowledge of in their tools. @ttmchiou, can we write a script to emulate what was done in https://github.com/tenstorrent/cloud/issues/2758 until we figure out the proper solution for this problem?

**Comment by ttmchiou** [comment]
yes, it should be possible to get a work-around for this script until we get a proper solution.

**Comment by milank94** [comment]
I ran into the same issue on a Galaxy 4U system (for a customer).

The order of the Coords by Bus ID was not in the configuration that the tt-metal was hardcoded for which caused failures. The workaround was to manually rebind the Bus IDs (swap 2 and 3).

![image](https://github.com/user-attachments/assets/f3984ec1-eaea-40ec-97e8-5dcad7b3a98f)

Ideally, if we can get a permanent fix for this from SysEng as we expect customers to be able to set these systems up without doing any temporary hacks.

Happy to share more details if needed or to push the priority of this item.

**Comment by TT-billteng** [comment]
> I ran into the same issue on a Galaxy 4U system (for a customer).
> 
> The order of the Coords by Bus ID was not in the configuration that the tt-metal was hardcoded for which caused failures. The workaround was to manually rebind the Bus IDs (swap 2 and 3).
> 
> ![image](https://private-user-images.githubusercontent.com/78221808/362951603-f3984ec1-eaea-40ec-97e8-5dcad7b3a98f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ5Nzc3MTQsIm5iZiI6MTcyNDk3NzQxNCwicGF0aCI6Ii83ODIyMTgwOC8zNjI5NTE2MDMtZjM5ODRlYzEtZWFlYS00MGVjLTk3ZTgtNWRjYWQ3YjNhOThmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODMwVDAwMjMzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMzMzU2YTU0NWZlMjliZjhhNjNiZWMyODc5ZGY4ODkyMjdlNDA4MjZmMjM4NjQxMmNhNTgzZDFjMjgxYzA0YzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._hs1bR2K79W42YiZb-lRFzayipmheczN5Q0nHly_5Fc)
> 
> Ideally, if we can get a permanent fix for this from SysEng as we expect customers to be able to set these systems up without doing any temporary hacks.
> 
> Happy to share more details if needed or to push the priority of this item.

The ideal solution I heard was to consume the output of `tt-topology` properly

**Comment by TT-billteng** [comment]
> @cfjchu - Hi Joseph, can you take a look? it's a P0 - thanks!

The proper solution is a little more involved and needs some cross-functional collaboration. In the meantime, @ttmchiou is writing a script to fix the device order of these T3Ks

**Comment by ttmchiou** [comment]
Script has been written and encountered some ND issues testing T3K VMs.
Running t3k unit tests shows failing tests (these tests do not fail on main branch).

Example test runs (some use repeated attempts)
https://github.com/tenstorrent/tt-metal/actions/runs/10692579064/job/29644656223
https://github.com/tenstorrent/tt-metal/actions/runs/10693593480/job/29699885419
https://github.com/tenstorrent/tt-metal/actions/runs/10710904828
https://github.com/tenstorrent/tt-metal/actions/runs/10723976170/job/29739148116

Reprovisioned t3k VMs 3 times to confirm not an issue with fw or drivers.

We currently have 1 t3k VM thats in operation (manually scripted) and Im using this as a basis to see if there is a difference.

t3k-04 has been noted to have a potential hardware issue*

**Comment by jvasilje** [comment]
@TT-billteng @cfjchu does this issue show up on bare metal T3Ks...?

Customer team has flagged this issue as a blocker to shipping T3ks - is that the case? 
When a customer buys a LoudBox or QuietBox - will they run into this issue?

Or does this only show up with fancy connections in the cloud towards the goal of enabling VMs?

**Comment by TT-billteng** [comment]
> @TT-billteng @cfjchu does this issue show up on bare metal T3Ks...?
> 
> Customer team has flagged this issue as a blocker to shipping T3ks - is that the case? When a customer buys a LoudBox or QuietBox - will they run into this issue?
> 
> Or does this only show up with fancy connections in the cloud towards the goal of enabling VMs?

seems like it does show up on bare metal T3Ks

@ttmchiou has submitted a script that can be used to fix the ordering of the devices, but this is just a workaround https://github.com/tenstorrent/tt-metal/commit/37bc0d3a91b9b5462fbf69f15070025d501f23a9

**Comment by mbahnasTT** [comment]
Thanks!

**Comment by cfjchu** [comment]
@TT-billteng @ttmchiou fix has been merged in so workaround is no longer needed.
